productcat/src/main/java/com/visenze/productcat/android/data/DataCollection.java

```diff
@@ -0,0 +1,88 @@
+package com.visenze.productcat.android.data;
+
+import android.Manifest;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.telephony.TelephonyManager;
+
+import androidx.core.app.ActivityCompat;
+
+import com.visenze.productcat.android.model.DeviceInfo;
+
+import java.io.UnsupportedEncodingException;
+
+public class DataCollection implements GetGAIDTask.OnTaskSuccess {
+
+    private GetGAIDTask gaidTask;
+
+    private String gaid;
+    private String didmd5;
+
+    public DataCollection(Context context) {
+
+        generateDoubleHashIMEI(context);
+
+        gaidTask = new GetGAIDTask(context, this);
+        gaidTask.execute();
+
+
+    }
+
+    private void generateDoubleHashIMEI(Context context) {
+        String imei = getIMEI(context);
+        didmd5 = MD5(imei);
+        didmd5 = MD5(didmd5);
+    }
+
+    private String getIMEI(Context context) {
+        String deviceUniqueIdentifier = null;
+        TelephonyManager tm = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
+        if (null != tm) {
+            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE) == PackageManager.PERMISSION_GRANTED) {
+                deviceUniqueIdentifier = tm.getDeviceId();
+            }
+        }
+        return deviceUniqueIdentifier;
+    }
+
+    private String MD5(String md5) {
```

Comment:
rename param to imei or just string. Should create a SecurityHelper with a static method (or create a SecurityService) and write unit test for this . This is a separate Responsibility and should be abstracted into another class.

Reply:
added SecurityHelper in utils package